src/shims/posix/fs.rs

```diff
@@ -1196,6 +1215,85 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+        }
+
+        let open_dir = this.machine.dir_handler.streams.get_mut(&dirp).ok_or_else(|| {
+            err_unsup_format!("the DIR pointer passed to readdir64 did not come from opendir")
+        })?;
+
+        let entry = match open_dir.read_dir.next() {
+            Some(Ok(dir_entry)) => {
+                let ino64_t_layout = this.libc_ty_layout("ino64_t")?;
+                let off64_t_layout = this.libc_ty_layout("off64_t")?;
+                let c_ushort_layout = this.libc_ty_layout("c_ushort")?;
+                let c_uchar_layout = this.libc_ty_layout("c_uchar")?;
+
+                // If the host is a Unix system, fill in the inode number with its real value.
+                // If not, use 0 as a fallback value.
+                #[cfg(unix)]
+                let ino = std::os::unix::fs::DirEntryExt::ino(&dir_entry);
+                #[cfg(not(unix))]
+                let ino = 0u64;
+
+                let file_type = this.file_type_to_d_type(dir_entry.file_type())?;
+
+                let imms = [
+                    immty_from_uint_checked(ino, ino64_t_layout)?, // d_ino
+                    immty_from_uint_checked(0u128, off64_t_layout)?, // d_off
+                    immty_from_uint_checked(0u128, c_ushort_layout)?, // d_reclen
+                    immty_from_int_checked(file_type, c_uchar_layout)?, // d_type
+                ];
+                let name = dir_entry.file_name();
+                let name_bytes = os_str_to_bytes(&name)?;
+                let name_offset = imms.iter()
+                    .map(|imm| imm.layout.size.bytes())
+                    .sum::<u64>();
+                let size = name_offset
+                    .checked_add(u64::try_from(name_bytes.len()).unwrap())
+                    .unwrap()
+                    .checked_add(1)
+                    .unwrap();
+
+                let entry = this.malloc(size, /*zero_init:*/ false, MiriMemoryKind::C)?;
+                let entry_layout = this.layout_of(this.tcx.mk_array(this.tcx.types.u8, size))?;
+                let entry_place = MPlaceTy::from_aligned_ptr(entry, entry_layout);
+                this.write_packed_immediates(&entry_place, &imms)?;
+
+                let name_ptr = entry.offset(Size::from_bytes(name_offset), this)?;
+                this.memory.write_bytes(name_ptr, name_bytes.iter().copied().chain(std::iter::once(0)))?;
+
+                entry
+            },
+            None => {
+                // end of stream: return NULL
+                Pointer::null()
+            }
+            Some(Err(e)) => {
+                this.set_last_error_from_io_error(e.kind())?;
+                Pointer::null()
+            }
+        };
+
+        let open_dir = this.machine.dir_handler.streams.get_mut(&dirp).unwrap();
+        let old_entry = std::mem::replace(&mut open_dir.entry, entry);
+        this.free(old_entry, MiriMemoryKind::C)?;
+
+        Ok(Scalar::from_maybe_pointer(entry, this))
+    }
+
     fn linux_readdir64_r(
```

Comment:
Is it even worth keeping this code? It is not used any more by the standard library, and implements a deprecated interface. Seems okay for me to remove it. (That might also make for a smaller diff.)

Reply:
I wasn't sure if miri is supposed to work on older stdlib versions.  If we don't care about that (or about people doing `libc::readdir64_r()` manually) I'm happy to delete the `_r()` version.

Reply:
We definitely don't care about older stdlib versions.

I am not sure if there *are* people directly calling `readdir64_r`, and we only barely implement enough of these functions anyway to support libstd... so I'd be okay with removing the old code, and then if someone complains that they needed it we can always add it back.